Fix wide-angle lens flare occlusion lag

audrow opened the pull request:
Fixes #3310 

This forces an update of the wide angle dummy camera, which updates the Ogre camera's view matrix. 

---

One way to check that this fixes the bug, is to use [this](https://github.com/audrow/gazebo-classic/blob/audrow/debug-lens-flare/worlds/lensflare_wideangle_cam.world) world file. Open it in Gazebo, and then move the only camera (`wide_angle_cameras_occluded_higher`) from it's starting height (`z=21`) to `z=11`, where it will have a wall between it and the light source. You can then move it back to `z=21`.

You should see lens flare when the light source is in the image and now lens flare when the light source is occluded.

scpeters committed:
Add model to illustrate lens flare lag issue

This adds a model with wide-angle cameras updated at 1 Hz
that demonstrate the bug reported in #3310 when moved
back and forth.

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

audrow committed:
Force update of wide angle dummy camera

Signed-off-by: Audrow Nash <audrow@intrinsic.ai>

iche033 commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -305,6 +305,7 @@ namespace gazebo
                 ignition::math::Pose3d(
                   Conversions::ConvertIgn(cam->getDerivedPosition()),
                   Conversions::ConvertIgn(quat)));
+            this->dataPtr->wideAngleDummyCamera->OgreCamera()->getParentSceneNode()->_update(true, true);
```
nit: wrap line to 80 chars

audrow replied:
Fixed in 1207519.

audrow committed:
Wrap at 80 characters

Signed-off-by: Audrow Nash <audrow@intrinsic.ai>

scpeters commented:
this fixes it for me! I've been tinkering with how to add an integration test for it, but if I don't figure it out in my next time box, then I'll just skip it and approve this

scpeters committed:
Merge branch 'gazebo11' into scpeters/lens-flare-occlusion-lag

scpeters committed:
Fix indentation in world file

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Existing test: get sensor by scoped name

The camera names aren't unique in this world,
so use the scoped name when getting sensors.

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Add test for occlusion lag

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Merge remote-tracking branch 'audrow/audrow/fix-lens-flare-delay' into scpeters/lens-flare-occlusion-lag

scpeters commented:
> this fixes it for me! I've been tinkering with how to add an integration test for it, but if I don't figure it out in my next time box, then I'll just skip it and approve this

I've gotten a test working in branch `scpeters/lens-flare-occlusion-lag`. I opened a pull request targeting this PR in https://github.com/audrow/gazebo-classic/pull/3, but it didn't start Jenkins jobs because it's targeting your fork. If the test looks ok, then please merge it into this PR with a merge commit and we can confirm that the tests pass in CI

audrow committed:
Merge pull request #3 from gazebosim/scpeters/lens-flare-occlusion-lag

Add test for lens flare occlusion lag

audrow committed:
Use expect near in image equality checks

Signed-off-by: Audrow Nash <audrow@intrinsic.ai>

scpeters committed:
Try to robustify image waiting

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters commented:
> > this fixes it for me! I've been tinkering with how to add an integration test for it, but if I don't figure it out in my next time box, then I'll just skip it and approve this
> 
> I've gotten a test working in branch `scpeters/lens-flare-occlusion-lag`. I opened a pull request targeting this PR in [audrow#3](https://github.com/audrow/gazebo-classic/pull/3), but it didn't start Jenkins jobs because it's targeting your fork. If the test looks ok, then please merge it into this PR with a merge commit and we can confirm that the tests pass in CI

the new test case passed on Ubuntu but failed in the homebrew build on [mac-four.bigsur](https://build.osrfoundation.org/computer/mac-four.bigsur). I noticed all the GUI tests failed on that run as well, so there may be an issue with the CI machine. I just pushed an improvement to the image waiting logic in https://github.com/gazebosim/gazebo-classic/pull/3325/commits/29dbe5a03994dd7fb17b541ba98a4caf9494c173 that hopefully is more robust. The new CI build is also running on [mac-four.bigsur](https://build.osrfoundation.org/computer/mac-four.bigsur), so we'll see if it helps

scpeters commented:
now with https://github.com/gazebosim/gazebo-classic/commit/29dbe5a03994dd7fb17b541ba98a4caf9494c173 it passed on homebrew but failed repeatedly on Ubuntu. I'll keep testing locally to see if I can reproduce the flakiness

scpeters committed:
Fix test flakiness with --lockstep

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters commented:
> now with [29dbe5a](https://github.com/gazebosim/gazebo-classic/commit/29dbe5a03994dd7fb17b541ba98a4caf9494c173) it passed on homebrew but failed repeatedly on Ubuntu. I'll keep testing locally to see if I can reproduce the flakiness

I was able to reproduce the issue locally, and it appears in my local testing that `--lockstep` ( https://github.com/gazebosim/gazebo-classic/pull/3325/commits/032247ef245b984d4bc4b9230b2382edc0028b4c ) fixes the sensor timing and test flakiness

scpeters commented:
test passes in both CI runs; merging
